app/controllers/repositories_controller.rb

```diff
@@ -41,22 +42,35 @@ class RepositoriesController < ApplicationController
   menu_item :settings, only: :edit
   default_search_scope :changesets
 
-  before_filter :find_repository, except: :edit
   before_filter :find_project, only: :edit
+  before_filter :find_repository, except: [:edit, :destroy]
+  before_filter :find_project_by_project_id, only: :destroy
   before_filter :authorize
   accept_key_auth :revisions
 
-  rescue_from Redmine::Scm::Adapters::CommandFailed, with: :show_error_command_failed
+  rescue_from OpenProject::Scm::Adapters::CommandFailed, with: :show_error_command_failed
 
   def edit
-    @repository = @project.repository
-    if !@repository
-      @repository = Repository.factory(params[:repository_scm])
-      @repository.project = @project if @repository
+
+    @repository = @project.repository || Repository.factory(params[:repository_scm])
+    if @repository.nil?
+      flash.now[:error] = l(:repository_factory_error)
     end
-    if request.post? && @repository
-      @repository.attributes = params[:repository]
-      @repository.save
+
+    @repository.project = @project
+
+    if request.post?
+      if true # TODO params[:managed_epository]
+        service = Scm::CreateManagedRepositoryService.new(@repository)
+        if service.call
+          flash.now[:notice] = l 'repositories.managed.request_successful'
+        else
+          flash.now[:error] = l service.rejected_reason
+        end
+      else
+        @repository.attributes = params[:repository]
```

Comment:
@oliverguenther FYI, Rails 4.1 will raise an `ArgumentError` if `params[:repository]` is `nil`:

    When assigning attributes, you must pass a hash as an argument.